core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -95,30 +107,84 @@ public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand
 
    @Override
    public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
+      Map<Object, Object> originalMap = command.getMap();
+      ConsistentHash ch = dm.getConsistentHash();
+      Address localAddress = rpcManager.getAddress();
       if (ctx.isOriginLocal()) {
-         Set<Address> primaryOwners = new HashSet<Address>(command.getAffectedKeys().size());
-         for (Object k : command.getAffectedKeys()) {
-            primaryOwners.add(cdl.getPrimaryOwner(k));
+         List<NotifyingFuture<Object>> futures = new ArrayList<NotifyingFuture<Object>>(
+               rpcManager.getMembers().size() - 1);
+         // TODO: if async we don't need to do futures...
+         RpcOptions options = rpcManager.getDefaultRpcOptions(isSynchronous(command));
+         for (Address member : rpcManager.getMembers()) {
```

Comment:
Hrmm I was under the assumption that putAll wanted to be most optimized for very large data sets to make loading initial data significantly faster.

Also while working on improvements for getAll implementation I am wondering how much more efficient using the wrapper collections are.  I am basically trading less memory usage for higher CPU utilization since we have to iterate over the entries more times.